tests/56-basic-iterate_syscalls.c

```diff
@@ -0,0 +1,84 @@
+/**
+ * Seccomp Library test program
+ *
+ * Copyright (c) 2020 Red Hat <gscrivan@redhat.com>
+ * Author: Giuseppe Scrivano <gscrivan@redhat.com>
+ */
+
+/*
+ * This library is free software; you can redistribute it and/or modify it
+ * under the terms of version 2.1 of the GNU Lesser General Public License as
+ * published by the Free Software Foundation.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library; if not, see <http://www.gnu.org/licenses>.
+ */
+
+#include <errno.h>
+#include <string.h>
+#include <stdlib.h>
+#include <stdio.h>
+
+#include <seccomp.h>
+
+unsigned int arch_list[] = {
+	SCMP_ARCH_NATIVE,
+	SCMP_ARCH_X86,
+	SCMP_ARCH_X86_64,
+	SCMP_ARCH_X32,
+	SCMP_ARCH_ARM,
+	SCMP_ARCH_AARCH64,
+	SCMP_ARCH_MIPS,
+	SCMP_ARCH_MIPS64,
+	SCMP_ARCH_MIPS64N32,
+	SCMP_ARCH_MIPSEL,
+	SCMP_ARCH_MIPSEL64,
+	SCMP_ARCH_MIPSEL64N32,
+	SCMP_ARCH_PPC,
+	SCMP_ARCH_PPC64,
+	SCMP_ARCH_PPC64LE,
+	SCMP_ARCH_S390,
+	SCMP_ARCH_S390X,
+	SCMP_ARCH_PARISC,
+	SCMP_ARCH_PARISC64,
+	SCMP_ARCH_RISCV64,
+	-1
+};
+
+static int test_arch(int arch, int init)
+{
+	int n, iter = 0;
+
+	for (iter = init; iter < init + 1000; iter++) {
```

Comment:
So just to make sure I understand this loop:

We're incrementing an index and seeing if that matches a valid syscall for this architecture.  If it is valid syscall, try to look up the syscall number (on line 64) given the name.

Is it okay that we're using the gperf table to resolve the number (in line 60) and the name (in line 64)?  I think it's okay, but I have a fear that we're using code under test to verify the code under test ;).

Reply:
it is testing whether the resolved name given a syscall number returns the same syscall number.

The only reason why this test exists is to stress the lookup table :)

I've used the heuristic `init` to `init + 1000` just to make sure the search doesn't stop  if there is a gap in the numbering, e.g. if there are defined syscalls like: 1, 3, 4... make sure it keeps trying and not stop on the missing `2`.

In the first version, this check was part of `arch-syscall-validate` where I could look up directly into the syscalls table, but now that it is moved to `tests/`, I cannot access it directly.